### PR TITLE
feat(pyspark): add support for PySpark 3.5

### PR DESCRIPTION
**Pull request by kszucs**
depends on #7950

### CODE REVIEW

File: ibis/backends/conftest.py [hunk 1]
```diff
@@ -4,13 +4,15 @@
...
 import itertools
+import operator
 from functools import cache
 from pathlib import Path
 from typing import TYPE_CHECKING, Any
 
 import _pytest
 import numpy as np
 import pandas as pd
+import pkg_resources
```
Comment (cpcloud) [inline ibis/backends/conftest.py:15]: Can you use `importlib.metadata` instead? `pkg_resources` is kind of awful.

File: pyproject.toml [hunk 2]
```diff
@@ -79,7 +79,7 @@ pydata-google-auth = { version = ">=1.4.0,<2", optional = true }
 pydruid = { version = ">=0.6.5,<1", optional = true, extras = ["sqlalchemy"] }
 pymysql = { version = ">=1,<2", optional = true }
 pyodbc = { version = ">=4.0.39,<6", optional = true }
-pyspark = { version = ">=3,<3.4", optional = true } # pyspark is heavily broken by numpy >=1.24 and pandas >=2
+pyspark = { version = ">=3", optional = true }
```
Comment (cpcloud) [inline pyproject.toml:82]: Should we keep an upper bound on this at `<4`? PySpark does have a history of being incompatible with the latest pandas and numpy.

**Review by cpcloud — APPROVED** [review]
Sweet!

File: ibis/backends/conftest.py (hunk 1 above)
Comment (kszucs) [reply ibis/backends/conftest.py:15]: Sure, updating.

File: pyproject.toml (hunk 2 above)
Comment (kszucs) [reply pyproject.toml:82]: Added.

**Review by cpcloud — APPROVED** [review]

**Pull request merged**